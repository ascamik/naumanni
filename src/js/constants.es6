export const NAUMANNI_VERSION = `naumanni ${process.env.NAUMANNI_VERSION || 'development ver.'}`

export const REGEX_PGP_FINGERPRINT = /PGP Key Fingerprint: ([0-9a-fA-F]+)/

export const COLUMN_TIMELINE = 'timeline'
export const COLUMN_FRIENDS = 'friends'
export const COLUMN_TALK = 'talk'

export const TIMELINE_FEDERATION = 'federation'
export const TIMELINE_LOCAL = 'local'
export const TIMELINE_HOME = 'home'

export const DIALOG_ADD_ACCOUNT = 'addAccount'
export const DIALOG_AUTHORIZE_ACCOUNT = 'authorizeAccount'

// MIXEDの方が良いかも
export const SUBJECT_MIXED = '!!__SUBJECT_MIXED__!!'   // Symbolだとjsonizeできないのでダメ

export const VISIBLITY_DIRECT = 'direct'
export const VISIBLITY_PRIVATE = 'private'
export const VISIBLITY_UNLISTED = 'unlisted'
export const VISIBLITY_PUBLIC = 'public'

export const NOTIFICATION_TYPE_MENTION = 'mention'
export const NOTIFICATION_TYPE_REBLOG = 'reblog'
export const NOTIFICATION_TYPE_FAVOURITE = 'favourite'
export const NOTIFICATION_TYPE_FOLLOW = 'follow'

export const KEY_ENTER = 13
