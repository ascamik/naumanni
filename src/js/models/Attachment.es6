import {Record} from 'immutable'


const AttachmentRecord = Record({  // eslint-disable-line new-cap
  'id': '',
  'type': '',
  'url': '',
  'remote_url': '',
  'preview_url': '',
  'text_url': '',
})


/**
 * Mastodon„ÅÆMedia
 */
export default class Attachment extends AttachmentRecord {
}
