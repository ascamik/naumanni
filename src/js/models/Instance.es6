/* @flow */
import {fromJS, Map, Record} from 'immutable'

import type OAuthToken from 'src/models/OAuthToken'

import {
  STREAM_HOME, STREAM_LOCAL, STREAM_FEDERATION, STREAM_TAG,
} from 'src/constants'

type Url = string
type InstanceShape = {
  uri: string,
  title: string,
  description: string,
  email: string,
  version: string,
  urls: Map<string, string>
};


const URL_KEY_STREAMING_API = 'streaming_api'
const instanceDefaults: InstanceShape = {
  host: '',
  uri: '',
  title: '',
  description: '',
  email: '',
  version: '',
  urls: new Map(),
}

/**
 * Mastodon Instance Information
 * see https://github.com/tootsuite/documentation/blob/master/Using-the-API/API.md#instance
 */
export default class Instance extends Record(instanceDefaults) {  // eslint-disable-line new-cap
  /**
   * @constructor
   * @param {string} host
   * @param {any} raw
   */
  constructor(host: string, raw: {urls?: {[string]: string}}) {
    super({
      ...raw,
      host,
      urls: fromJS(raw.urls || {}),
    })
  }

  /**
   * WSSのURLを返す
   * @param {OAuthToken} token
   * @param {string} stream
   * @param {object} params
   * @return {Url}
   */
  makeStreamingAPIUrl(token: OAuthToken, stream: string, params?: any={}): Url {
    require('assert')(
      [STREAM_FEDERATION, STREAM_TAG, STREAM_LOCAL, STREAM_HOME].indexOf(stream) >= 0,
      'valid stream type'
    )

    let rootUrl = this.urls.get(URL_KEY_STREAMING_API)
    if(!rootUrl) {
      rootUrl = `wss://${token.host}`
    }

    const otherParams = Object.keys(params).length === 0
      ? ''
      : Object.keys(params)
          .reduce((prev, k, i) => (
            i === 0
              ? `&${k}=${encodeURIComponent(params[k])}`
              : `${prev}&${k}=${encodeURIComponent(params[k])}`
          ), '')

    return `${rootUrl}/api/v1/streaming/?access_token=${token.accessToken}&stream=${stream}${otherParams}`
  }
}
