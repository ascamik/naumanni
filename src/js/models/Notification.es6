import {Record} from 'immutable'
import moment from 'moment'


const NotificationRecord = Record({  // eslint-disable-line new-cap
  host: '',

  id: null,
  type: '',
  created_at: '',
  account: null,
  status: null,
})


/**
 * Mastodon's Notification
 */
export default class Notification extends NotificationRecord {
  get uri() {
    return `notification:${this.host}:${this.id}`
  }

  get createdAt() {
    return moment(this.created_at)
  }

  // TODO: どっかに纏める
  static compareCreatedAt(a, b) {
    const aAt = a.createdAt
    const bAt = b.createdAt
    if(aAt.isBefore(bAt))
      return 1
    else if(aAt.isAfter(bAt))
      return -1
    return 0
  }
}
