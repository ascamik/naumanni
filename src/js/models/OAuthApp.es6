import {Record} from 'immutable'


const OAuthAppRecord = Record({  // eslint-disable-line new-cap
  host: '',
  id: '',
  redirect_uri: '',
  client_id: '',
  client_secret: '',
})


/**
 * OAuthでのApp登録を表すModel
 */
export default class OAuthApp extends OAuthAppRecord {
  static storeName = 'oauth_apps'
  static keyPath = 'address'
  static indexes = [
    ['host', 'host', {}],  // uniqueでも良いかも
    ['client_id', ['host', 'client_id'], {unique: true}],
  ]

  /**
   * @constructor
   */
  constructor(...args) {
    super(...args)
  }

  /**
   * @override
   */
  toJS() {
    const js = super.toJS()

    js.address = this.address
    return js
  }

  /**
   * 一意な識別子を返す
   */
  get address() {
    return `${this.host}::${this.id}`
  }
}
