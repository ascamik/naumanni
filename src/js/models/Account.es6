import {Record} from 'immutable'
import {REGEX_PGP_FINGERPRINT} from 'src/constants'

const AccountRecord = Record({  // eslint-disable-line new-cap
  host: '',

  id: null,
  username: null,
  acct: null,
  display_name: null,
  locked: null,
  created_at: null,
  followers_count: null,
  following_count: null,
  statuses_count: null,
  note: null,
  url: null,
  avatar: null,
  avatar_static: null,
  header: null,
  header_static: null,
})


/**
 * OAuthでのApp登録を表すModel
 */
export default class Account extends AccountRecord {
  static storeName = 'accounts'
  static keyPath = 'address'
  static indexes = [
    ['host', 'host', {}],
  ]

  /**
   * @constructor
   */
  constructor(...args) {
    super(...args)
  }

  /**
   * @override
   */
  toJS() {
    const js = super.toJS()

    js.address = this.address
    return js
  }

  /**
   * 一意な識別子を返す
   */
  get address() {
    return `${this.id}@${this.host}`
  }

  get account() {
    if(this.acct.indexOf('@') >= 0)
      return this.acct
    return `${this.acct}@${this.host}`
  }

  get hasPublicKey() {
    return this.note.match(REGEX_PGP_FINGERPRINT) ? true : false
  }

  get publicKeyId() {
    const match = this.note.match(REGEX_PGP_FINGERPRINT)
    if(!match)
      return null
    const fingerprint = match[1]
    if(fingerprint.length != 40)
      return null
    return fingerprint.substring(24)
  }

  get privateKeyArmored() {
    return localStorage.getItem(`pgp::privateKey::${this.address}`)
  }
}
