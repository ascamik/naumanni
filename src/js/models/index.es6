import Account from './Account'
import Attachment from './Attachment'
import OAuthApp from './OAuthApp'
import OAuthToken from './OAuthToken'
import Status from './Status'
export {Account, OAuthApp, OAuthToken, Status}

import UIColumn from './UIColumn'
export {UIColumn}
