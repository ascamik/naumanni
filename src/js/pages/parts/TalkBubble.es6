/* @flow */
import React from 'react'
import classNames from 'classnames'
import {List} from 'immutable'
import {FormattedDate} from 'react-intl'

import {Attachment} from 'src/models'
import {IconFont, SafeContent} from './'


type Props = {
  createdAt: Date,
  isEncrypted: boolean,
  mediaFiles: List<Attachment>,
  parsedContent: List<{[type: string]: {[string]: any}}>,
  onClickHashTag: (string, SyntheticEvent) => void,
  onClickMedia: (e: SyntheticEvent, files: List<Attachment>, idx: number) => void,
}

export default class TalkBubble extends React.PureComponent {
  props: Props

  render() {
    const {createdAt, isEncrypted, mediaFiles, parsedContent, onClickHashTag} = this.props

    return (
      <li>
        <div className={`status-content ${isEncrypted ? 'is-encrypted' : ''}`}>
          <SafeContent parsedContent={parsedContent} onClickHashTag={onClickHashTag} />
          {this.renderMediaFilesInTalk(mediaFiles)}
        </div>
        <div className="status-date">
          <FormattedDate value={createdAt}
            year="numeric" month="2-digit" day="2-digit"
            hour="2-digit" minute="2-digit" second="2-digit"
          />
        </div>
        {isEncrypted && <div className="status-isEncrypted"><IconFont iconName="lock" /></div>}
      </li>
    )
  }

  renderMediaFilesInTalk(files: List<Attachment>) {
    if(!files.size)
      return null

    return (
      <div className={classNames(
        'status-mediaList',
        `status-mediaList${files.size}`,
      )}>
        {files.map((media, idx) => (
          <a key={media.preview_url}
            className="status-media"
            style={{backgroundImage: `url(${media.preview_url})`}}
            target="_blank"
            href={media.url}
            onClick={(e) => this.props.onClickMedia(e, files, idx)}
            />
        ))}
      </div>
    )
  }
}
