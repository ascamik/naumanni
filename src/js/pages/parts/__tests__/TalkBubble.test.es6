import React from 'react'
import {mountWithIntl} from 'enzyme-react-intl'
import serializer from 'enzyme-to-json/serializer'
import {List} from 'immutable'

import {Attachment} from 'src/models'
import {TOKEN_HASHTAG} from 'src/constants'
import TalkBubble from '../TalkBubble'


expect.addSnapshotSerializer(serializer)

describe('TalkBubble', () => {
  let defaultProps = {
    createdAt: new Date(2017, 7, 28),
    isEncrypted: false,
    mediaFiles: List([
      new Attachment({
        url: 'https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg',
        preview_url: 'https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg',
      })
    ]),
    parsedContent: List([
      {
        type: TOKEN_HASHTAG,
        tag: 'naumanni',
        url: 'https://friends.nico/tags/naumanni',
      }
    ]),
    onClickHashTag: jest.fn(),
    onClickMedia: jest.fn(),
  }

  describe('snapshot', () => {
    it('isEncrypted == false', () => {
      const wrapper = mountWithIntl(<TalkBubble {...defaultProps} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('isEncrypted == true', () => {
      const props = {
        ...defaultProps,
        isEncrypted: true,
      }
      const wrapper = mountWithIntl(<TalkBubble {...props} />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  it('propagate hash tag link click event', () => {
    const onClickHashTag = jest.fn()
    const props = {
      ...defaultProps,
      onClickHashTag,
    }
    const wrapper = mountWithIntl(<TalkBubble {...props} />)

    const hashtag = wrapper.find('p > a')
    hashtag.simulate('click')

    expect(onClickHashTag).toHaveBeenCalled()
  })

  it('propagate media click event', () => {
    const onClickMedia = jest.fn()
    const props = {
      ...defaultProps,
      onClickMedia,
    }
    const wrapper = mountWithIntl(<TalkBubble {...props} />)

    const media = wrapper.find('a.status-media')
    media.simulate('click')

    expect(onClickMedia).toHaveBeenCalled()
  })
})
