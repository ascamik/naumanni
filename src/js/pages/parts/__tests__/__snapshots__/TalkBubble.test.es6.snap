// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TalkBubble snapshot isEncrypted == false 1`] = `
<TalkBubble
  createdAt={2017-08-27T15:00:00.000Z}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "textComponent": "span",
    }
  }
  isEncrypted={false}
  mediaFiles={
    Array [
      Object {
        "id_by_host": Object {},
        "meta": undefined,
        "preview_url": "https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg",
        "preview_url_by_host": Object {},
        "text_url": "",
        "type": "",
        "url": "https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg",
        "url_by_host": Object {},
      },
    ]
  }
  onClickHashTag={[Function]}
  onClickMedia={[Function]}
  parsedContent={
    Array [
      Object {
        "tag": "naumanni",
        "type": "hashtag",
        "url": "https://friends.nico/tags/naumanni",
      },
    ]
  }
>
  <li>
    <div
      className="status-content "
    >
      <SafeContent
        onClickHashTag={[Function]}
        parsedContent={
          Array [
            Object {
              "tag": "naumanni",
              "type": "hashtag",
              "url": "https://friends.nico/tags/naumanni",
            },
          ]
        }
      >
        <p>
          <a
            onClick={[Function]}
          >
            #
            naumanni
          </a>
        </p>
      </SafeContent>
      <div
        className="status-mediaList status-mediaList1"
      >
        <a
          className="status-media"
          href="https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg"
          onClick={[Function]}
          style={
            Object {
              "backgroundImage": "url(https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg)",
            }
          }
          target="_blank"
        />
      </div>
    </div>
    <div
      className="status-date"
    >
      <FormattedDate
        day="2-digit"
        hour="2-digit"
        minute="2-digit"
        month="2-digit"
        second="2-digit"
        value={2017-08-27T15:00:00.000Z}
        year="numeric"
      >
        <span>
          08/28/2017, 12:00:00 AM
        </span>
      </FormattedDate>
    </div>
  </li>
</TalkBubble>
`;

exports[`TalkBubble snapshot isEncrypted == true 1`] = `
<TalkBubble
  createdAt={2017-08-27T15:00:00.000Z}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "textComponent": "span",
    }
  }
  isEncrypted={true}
  mediaFiles={
    Array [
      Object {
        "id_by_host": Object {},
        "meta": undefined,
        "preview_url": "https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg",
        "preview_url_by_host": Object {},
        "text_url": "",
        "type": "",
        "url": "https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg",
        "url_by_host": Object {},
      },
    ]
  }
  onClickHashTag={[Function]}
  onClickMedia={[Function]}
  parsedContent={
    Array [
      Object {
        "tag": "naumanni",
        "type": "hashtag",
        "url": "https://friends.nico/tags/naumanni",
      },
    ]
  }
>
  <li>
    <div
      className="status-content is-encrypted"
    >
      <SafeContent
        onClickHashTag={[Function]}
        parsedContent={
          Array [
            Object {
              "tag": "naumanni",
              "type": "hashtag",
              "url": "https://friends.nico/tags/naumanni",
            },
          ]
        }
      >
        <p>
          <a
            onClick={[Function]}
          >
            #
            naumanni
          </a>
        </p>
      </SafeContent>
      <div
        className="status-mediaList status-mediaList1"
      >
        <a
          className="status-media"
          href="https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg"
          onClick={[Function]}
          style={
            Object {
              "backgroundImage": "url(https://d2zoeobnny43zx.cloudfront.net/media_attachments/files/000/904/383/small/d8ffd22736103a82.jpg)",
            }
          }
          target="_blank"
        />
      </div>
    </div>
    <div
      className="status-date"
    >
      <FormattedDate
        day="2-digit"
        hour="2-digit"
        minute="2-digit"
        month="2-digit"
        second="2-digit"
        value={2017-08-27T15:00:00.000Z}
        year="numeric"
      >
        <span>
          08/28/2017, 12:00:00 AM
        </span>
      </FormattedDate>
    </div>
    <div
      className="status-isEncrypted"
    >
      <IconFont
        iconName="lock"
      >
        <span
          className="icon icon-lock"
        />
      </IconFont>
    </div>
  </li>
</TalkBubble>
`;
